Panic when using "over" in an expression context

Repro is with Zed commit f1be6a4.

@mattnibs and I were working toward responding to a community user's inquiry in a [Slack thread](https://brimdata.slack.com/archives/CTT24051C/p1710456812385789). I'd gotten to the point where I had this working in a user-defined operator:

```
$ cat op.zed 
op MyFlatten(val): (
  yield (over val | over this | collect(this))

$ zq -version
Version: v1.14.0-17-gf1be6a4a

$ echo '[[1,2],[3,4]]' | zq -I op.zed 'MyFlatten(this)' -
[1,2,3,4]
```

So I thought to try the equivalent in a user-defined function, which triggered this panic.

```
$ cat func.zed 
func MyFlatten(val): (
  (over val | over this | collect(this))
)

$ echo '[[1,2],[3,4]]' | zq -I func.zed 'yield MyFlatten(this)' -
panic: runtime error: index out of range [0] with length 0

goroutine 53 [running]:
github.com/brimdata/zed/runtime/sam/expr.Var.Eval(...)
	/Users/phil/work/zed/runtime/sam/expr/var.go:14
github.com/brimdata/zed/runtime/sam/op/traverse.(*Over).over(0xc0003e3da0, {0x28d5478?, 0xc000789f50}, {{0x28c76d0?, 0x36de9e0?}, 0x0?, 0x0?})
	/Users/phil/work/zed/runtime/sam/op/traverse/over.go:72 +0x164
github.com/brimdata/zed/runtime/sam/op/traverse.(*Over).Pull(0xc0003e3da0, 0xc0?)
	/Users/phil/work/zed/runtime/sam/op/traverse/over.go:56 +0x1a9
github.com/brimdata/zed/runtime/sam/op/traverse.(*Scope).run(0xc0006d1c70)
	/Users/phil/work/zed/runtime/sam/op/traverse/scope.go:69 +0x37
created by github.com/brimdata/zed/runtime/sam/op/traverse.(*Scope).Pull.func1 in goroutine 52
	/Users/phil/work/zed/runtime/sam/op/traverse/scope.go:47 +0x56
```

Meanwhile @mattnibs was coming up with a more sophisticated variant that was closer to the user's original intent, and it looks like he hit a similar panic.

```
$ cat maybe_flatten.zed
func maybe_flatten_fn(a): (
   kind(a) == "array"  
     ? (over a | and(kind(this) == "array")) 
        ? "It's an array of arrays, so I would flatten"
        : "It's an array containing some non-array values, so I would not flatten"
     : "It's not even an array, so I ain't touchin' that"
)

yield [[1,2],[3,4]], "hello"
| yield maybe_flatten_fn(this)

$ echo null | zq -I maybe_flatten.zed -
panic: runtime error: index out of range [0] with length 0

goroutine 25 [running]:
github.com/brimdata/zed/runtime/sam/expr.Var.Eval(...)
	/Users/phil/work/zed/runtime/sam/expr/var.go:14
github.com/brimdata/zed/runtime/sam/op/traverse.(*Over).over(0xc00056e840, {0x28d5478?, 0xc000876e10}, {{0x28c76d0?, 0x36de9e0?}, 0x0?, 0x0?})
	/Users/phil/work/zed/runtime/sam/op/traverse/over.go:72 +0x164
github.com/brimdata/zed/runtime/sam/op/traverse.(*Over).Pull(0xc00056e840, 0x0?)
	/Users/phil/work/zed/runtime/sam/op/traverse/over.go:56 +0x1a9
github.com/brimdata/zed/runtime/sam/op/traverse.(*Scope).run(0xc0006788c0)
	/Users/phil/work/zed/runtime/sam/op/traverse/scope.go:69 +0x37
created by github.com/brimdata/zed/runtime/sam/op/traverse.(*Scope).Pull.func1 in goroutine 24
	/Users/phil/work/zed/runtime/sam/op/traverse/scope.go:47 +0x56
```

As a side-effect of working on this, we also both recognized that this functionality should be mentioned in the [expressions docs](https://zed.brimdata.io/docs/language/expressions), but we might as well fix the bug before calling more attention to it.

--- a/runtime/sam/op/traverse/eval.go
+++ b/runtime/sam/op/traverse/eval.go
@@ -42,8 +42,10 @@ func (e *Expr) SetExit(exit *Exit) {
 }
 
 func (e *Expr) Eval(ectx expr.Context, this zed.Value) zed.Value {
+	b := zbuf.NewArray([]zed.Value{this})
+	b.SetVars(ectx.Vars())
 	select {
-	case e.batchCh <- zbuf.NewArray([]zed.Value{this}):
+	case e.batchCh <- b:
 	case <-e.ctx.Done():
 		return e.zctx.NewError(e.ctx.Err())
 	}

--- a/runtime/sam/op/traverse/ztests/over-expr-outer-vars.yaml
+++ b/runtime/sam/op/traverse/ztests/over-expr-outer-vars.yaml
@@ -1,0 +1,11 @@
+zed: |
+  func sumvals(vals): (
+    (over vals | sum(this))
+  )
+  sum := sumvals(v)
+
+input: |
+  {v:[1,2,3,4]}
+
+output: |
+  {v:[1,2,3,4],sum:10}

--- a/zbuf/array.go
+++ b/zbuf/array.go
@@ -9,6 +9,7 @@ import (
 // the Reader interfaces.
 type Array struct {
 	values []zed.Value
+	vars   []zed.Value
 }
 
 var _ Batch = (*Array)(nil)
@@ -37,9 +38,12 @@ func (a *Array) Append(r zed.Value) {
 	a.values = append(a.values, r)
 }
 
+func (a *Array) SetVars(vars []zed.Value) {
+	a.vars = vars
+}
+
 func (a *Array) Vars() []zed.Value {
-	// XXX TBD
-	return nil
+	return a.vars
 }
 
 func (a *Array) Write(r zed.Value) error {

